File: web_app/frontend/src/types/mission.ts
import type {
  MissionStatus,
  MissionResponse,
  MissionGoalResponse,
  MissionDetailResponse,
} from './api';

export interface Goal {
  id: string;
  missionId: string;
  sequenceNo: number;
  frameId: string;
  position: {
    x: number;
    y: number;
    z: number;
  };
  orientation: {
    x: number;
    y: number;
    z: number;
    w: number;
  };
  reachedAt: string | null;
  createdAt: string;
}

export interface Mission {
  id: string;
  robotId?: string;
  name: string;
  description?: string | null;
  status: MissionStatus;
  createdBy?: string;
  startedAt?: string | null;
  completedAt?: string | null;
  failureReason?: string | null;
  createdAt?: string;
  updatedAt?: string;
  goals?: Goal[];
}

export function mapGoalResponseToGoal(response: MissionGoalResponse): Goal {
  return {
    id: response.id,
    missionId: response.mission_id,
    sequenceNo: response.sequence_no,
    frameId: response.frame_id,
    position: {
      x: response.position_x,
      y: response.position_y,
      z: response.position_z,
    },
    orientation: {
      x: response.orientation_x,
      y: response.orientation_y,
      z: response.orientation_z,
      w: response.orientation_w,
    },
    reachedAt: response.reached_at,
    createdAt: response.created_at,
  };
}

export function mapMissionResponseToMission(
  response: MissionResponse | MissionDetailResponse,
): Mission {
  const goals =
    'goals' in response && Array.isArray(response.goals)
      ? response.goals.map(mapGoalResponseToGoal)
      : undefined;

  return {
    id: response.id,
    robotId: response.robot_id,
    name: response.name,
    description: response.description,
    status: response.status,
    createdBy: response.created_by,
    startedAt: response.started_at,
    completedAt: response.completed_at,
    failureReason: response.failure_reason,
    createdAt: response.created_at,
    updatedAt: response.updated_at,
    goals,
  };
}

File: web_app/backend/app/schemas/mission.py
import math
from datetime import datetime
from uuid import UUID

from pydantic import Field, field_validator, model_validator

from app.core.constants import MissionStatus
from app.schemas.common import APIModel, require_aware


class MissionGoalCreate(APIModel):
    sequence_no: int | None = Field(default=None, ge=0)
    frame_id: str = Field(default="map", min_length=1, max_length=128)
    position_x: float
    position_y: float
    position_z: float = 0.0
    orientation_x: float = 0.0
    orientation_y: float = 0.0
    orientation_z: float
    orientation_w: float

    @model_validator(mode="after")
    def validate_pose(self) -> "MissionGoalCreate":
        values = (
            self.position_x,
            self.position_y,
            self.position_z,
            self.orientation_x,
            self.orientation_y,
            self.orientation_z,
            self.orientation_w,
        )
        if not all(math.isfinite(value) for value in values):
            raise ValueError("pose values must be finite")
        norm = math.sqrt(
            self.orientation_x**2
            + self.orientation_y**2
            + self.orientation_z**2
            + self.orientation_w**2
        )
        if not math.isclose(norm, 1.0, rel_tol=1e-4, abs_tol=1e-4):
            raise ValueError("orientation quaternion must be normalized")
        return self


class MissionGoalResponse(APIModel):
    id: UUID
    mission_id: UUID
    sequence_no: int
    frame_id: str
    position_x: float
    position_y: float
    position_z: float
    orientation_x: float
    orientation_y: float
    orientation_z: float
    orientation_w: float
    reached_at: datetime | None = None
    created_at: datetime

    _aware_reached = field_validator("reached_at")(require_aware)
    _aware_created = field_validator("created_at")(require_aware)


class MissionCreate(APIModel):
    robot_id: UUID
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    goals: list[MissionGoalCreate] = Field(default_factory=list, max_length=500)


class MissionUpdate(APIModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    status: MissionStatus | None = None
    failure_reason: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def not_empty(self) -> "MissionUpdate":
        if not self.model_fields_set:
            raise ValueError("at least one field must be supplied")
        return self


class MissionResponse(APIModel):
    id: UUID
    robot_id: UUID
    name: str
    description: str | None = None
    status: MissionStatus
    created_by: UUID
    started_at: datetime | None = None
    completed_at: datetime | None = None
    failure_reason: str | None = None
    created_at: datetime
    updated_at: datetime

    _aware_started = field_validator("started_at")(require_aware)
    _aware_completed = field_validator("completed_at")(require_aware)
    _aware_created = field_validator("created_at")(require_aware)
    _aware_updated = field_validator("updated_at")(require_aware)


class MissionDetailResponse(MissionResponse):
    goals: list[MissionGoalResponse] = Field(default_factory=list)
